### PR TITLE
Add  `--kubernetes` flag to make dapr invoke cmd support kubernetes

### PR DESCRIPTION
**Pull request by imneov**
# Description

Support the invoke cmd on kubernetes mode. 

The process is similar:
1. Get the Pod corresponding to the App
2. Complete the specific communication through the `SubResource("proxy")` of `RESTClient`
3. Direct access to App's `AppPort` port or access to daprd's `HttpPort` port

```

                                      +-----------------+
                                      |                 |
                     AppPort          | +-------------+ |
         +----------------------------> |    APP      | |
  +----------+                        | +-------------+ |
  | dapr cli |                        |                 |
  +----------+                        | +-------------+ |
         +----------------------------> |    Daprd    | |
                    HttpPort          | +-------------+ |
                                      |                 |
                                      |            pod  |
                                      +-----------------+

```

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[[436](https://github.com/dapr/cli/issues/436)]_ #_[[848](https://github.com/dapr/cli/issues/848)]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

**Review by tanvigour — COMMENTED** [review]
Can we add test case for this?

**Comment by imneov** [comment]
> Can we add test case for this?

Sorry, just saw the reply.

Thank you tanvigour for review. I wrote part of the test code. There may be problems with code style or edge cases. Please review.

Welcome to add test cases.

**Comment by imneov** [comment]
There is a problem with using `proxy` to access the `HttpPort`, mainly because the address that daprd listens to is `127.0.0.1`, and the proxy cannot access this port.

So now our project(https://github.com/tkeel-io/cli/issues/74) uses `portforward` for `invoke`. The basic process is as follows:

1. Get the Pod corresponding to the App
2. Build a tunnel between the local and target pod through RESTClient's SubResource("portforward")
3. Access daprd's `HttpPort` port through the local tunnel port

This part of the implementation needs to modify `portforward.go` to implement random ports (https://github.com/dapr/cli/issues/863).

**Comment by mukundansundar** [comment]
@imneov Can you signoff on your commits? https://github.com/dapr/cli/blob/master/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work

**Comment by mukundansundar** [comment]
@imneov Can you fix conflicts in the PR?

**Comment by imneov** [comment]
> @imneov Can you fix conflicts in the PR?

Yes, I fix it.

File: cmd/invoke.go [hunk 1]
```diff
@@ -39,15 +40,18 @@ var (
...
-	Short: "Invoke a method on a given Dapr application. Supported platforms: Self-hosted",
+	Short: "Invoke a method on a given Dapr application. Supported platforms: Kubernetes and self-hosted",
 	Example: `
-# Invoke a sample method on target app with POST Verb
+# Invoke a sample method on target app with POST Verb in self-hosted mode
 dapr invoke --app-id target --method sample --data '{"key":"value"}
 
-# Invoke a sample method on target app with GET Verb
+# Invoke a sample method on target app with in Kubernetes
+dapr invoke -k --app-id target --method sample --data '{"key":"value"}
```
Comment (shubham1172) [inline cmd/invoke.go:49]: `--data` has an unclosed single quote. Same for the above example in self-hosted mode.

File: cmd/invoke.go [hunk 2]
```diff
@@ -79,11 +82,22 @@ dapr invoke --unix-domain-socket --app-id target --method sample --verb GET
 			}
 		}
 
-		response, err := client.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb, invokeSocket)
-		if err != nil {
-			err = fmt.Errorf("error invoking app %s: %w", invokeAppID, err)
-			print.FailureStatusEvent(os.Stderr, err.Error())
-			os.Exit(1)
+		var response string
+		if kubernetesMode {
```
Comment (shubham1172) [inline cmd/invoke.go:86]: Given that both kubernetes mode and standalone mode have same error handling, can we follow DRY and do this instead?
```go
if kubernetesMode {
    response, err = kubernetes.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb)
} else {
    client := standalone.NewClient()
    response, err = client.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb, invokeSocket)
}

// now handle error and response
if err != nil {
    // handle error
}
```

File: cmd/invoke.go (hunk 1 above)
Comment (imneov) [reply cmd/invoke.go]: Thank you for review. I fixed it.

File: cmd/invoke.go (hunk 2 above)
Comment (imneov) [reply cmd/invoke.go:87]: Thank you for review. I fixed it.

File: cmd/invoke.go [hunk 3]
```diff
@@ -79,7 +83,13 @@ dapr invoke --unix-domain-socket --app-id target --method sample --verb GET
...
 		}
 
-		response, err := client.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb, invokeSocket)
+		var response string
+		if kubernetesMode {
+			response, err = kubernetes.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb)
+		} else {
+			client := standalone.NewClient()
+			response, err = client.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb, invokeSocket)
+		}
```
Comment (shubham1172) [inline cmd/invoke.go:92]: nit: line break

Comment (imneov) [reply cmd/invoke.go:92]: How can I fix this?
`make lint` without any hints.

Comment (shubham1172) [reply cmd/invoke.go:92]: Oh you can simply add a new line after line 92. It's only for readability purpose.

Comment (imneov) [reply cmd/invoke.go:92]: OK, I fixed it.

File: pkg/kubernetes/invoke.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+func getAppInfoFromPod(p *DaprPod) (a *AppInfo) {
+	for _, c := range p.Spec.Containers {
+		if c.Name == "daprd" {
+			a = &AppInfo{
+				PodName:   p.Name,
+				Namespace: p.Namespace,
+			}
+			for i, arg := range c.Args {
+				if arg == "--app-port" {
```
[the remark this answers is not in the document]
Comment (imneov) [reply pkg/kubernetes/invoke.go:135]: I agree with this, but the style is the same as
https://github.com/dapr/cli/blob/4765d02e5ab4675985aa59cbc6d5ecd91ff14b68/pkg/kubernetes/list.go#L49-L55

File: cmd/invoke.go [hunk 5]
```diff
@@ -39,15 +40,19 @@ var (
...
 	Example: `
-# Invoke a sample method on target app with POST Verb
-dapr invoke --app-id target --method sample --data '{"key":"value"}
+# Invoke a sample method on target app with POST Verb in self-hosted mode
+dapr invoke --app-id target --method sample --data '{"key":"value"}'
 
-# Invoke a sample method on target app with GET Verb
+# Invoke a sample method on target app with in Kubernetes
+dapr invoke -k --app-id target --method sample --data '{"key":"value"}'
+
```
Comment (shubham1172) [inline cmd/invoke.go:50]: nit: remove extra line.

Comment (shubham1172) [inline cmd/invoke.go:50]: nit: remove extra line.

**Comment by mukundansundar** [comment]
@imneov Can you fix the DCO check error please?

**Comment by imneov** [comment]
> @imneov Can you fix the DCO check error please?

Done!

**Comment by mukundansundar** [comment]
@pravinpushkar Please review this PR

**Comment by pravinpushkar** [comment]
> @pravinpushkar Please review this PR

Looking at it now.

**Comment by pravinpushkar** [comment]
@imneov Please fix the small lint issue. PR looks fine.

**Comment by pravinpushkar** [comment]
ping @imneov

**Comment by imneov** [comment]
> @imneov Please fix the small lint issue. PR looks fine. We need to update this here also to be in sync - https://docs.dapr.io/reference/cli/dapr-invoke/

Do I need to submit a PR for docs?

**Comment by shubham1172** [comment]
> > > @imneov Please fix the small lint issue. PR looks fine. We need to update this here also to be in sync - https://docs.dapr.io/reference/cli/dapr-invoke/
> 
> @pravinpushkar Sorry I didn't find the lint error. what command do I need to run?

The lint error is available in this view https://github.com/dapr/cli/pull/862/files
image

(there is only one of such).

**Comment by imneov** [comment]
@pravinpushkar HI~  Is there anything else I need to do?

**Comment by pravinpushkar** [comment]
> > @imneov Please fix the small lint issue. PR looks fine. We need to update this here also to be in sync - https://docs.dapr.io/reference/cli/dapr-invoke/
> 
> Do I need to submit a PR for docs?

Yes, If not done already then please create a doc issue and PR and link it here.
I have triggered the PR checks let's see.

**Comment by imneov** [comment]
> Support the invoke cmd on kubernetes mode.

OK , Thanks for you reply.   @pravinpushkar 

There is  doc issue(https://github.com/dapr/docs/issues/2846) and PR(https://github.com/dapr/docs/pull/2847)

**Comment by imneov** [comment]
@pravinpushkar Hi~ Is there any work I need to do?

**Review by mukundansundar — COMMENTED** [review]
please add E2E for this feature.

**Comment by msfussell** [comment]
@mukundansundar - What is happening with this issue/PR?

**Comment by cicoyle** [comment]
@imneov - looks like there are some e2e failures, mind looking into and fixing those?
@dapr/maintainers-cli - looks like this PR has been open quite a while, we could use a review on this please :)

That being said, given the time constraint for this release, I will push this to the next release

**Comment by imneov** [comment]
Rebased onto current master as a single clean commit, added a dedicated kubernetes E2E and fixed two real CI blockers on the latest run:

1. `pkg/kubernetes/invoke.go` — converted the `else-if` chain in `getAppInfoFromPod` to a tagged `switch`. `staticcheck QF1003` is a hard failure under `golangci-lint v2.10.1`, which is what was breaking `Build linux_amd64 binaries`. This is the lint @pravinpushkar flagged a while back.

2. `tests/e2e/kubernetes/invoke_test.go` (new) — end-to-end coverage for `dapr invoke -k`: deploys a daprized nginx pod, waits for the sidecar, runs `dapr invoke -k --method / --verb GET` and asserts the response. Also covers the "app not found" negative path. The first run actually passed all four body subtests; the parent failed because I was calling `t.Run` from inside `t.Cleanup`, which Go forbids — moved teardown into a `defer` where `t.Run` is allowed.

The remaining failing jobs (`Upgrade path tests` HA, `Self-hosted darwin`) are flaky on master itself — last 20 scheduled runs on master show ~50% and ~35% failure rates respectively, with success/failure alternating minutes apart. They need a rerun rather than a fix.

Could one of @cicoyle / @pravinpushkar / @mukundansundar take another look when you have a moment? The docs counterpart (dapr/docs#2847) was auto-closed while this sat — I'll re-open a fresh one as soon as this merges.
